gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFileInputSource.java

```diff
@@ -0,0 +1,258 @@
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.apache.hadoop.mapreduce.TaskAttemptID;
+import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
+import org.apache.hadoop.mapreduce.lib.input.FileSplit;
+import org.apache.hadoop.util.ReflectionUtils;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.io.BaseEncoding;
+import com.google.common.io.Closer;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.SourceState;
+import gobblin.configuration.State;
+import gobblin.configuration.WorkUnitState;
+import gobblin.source.Source;
+import gobblin.source.extractor.Extractor;
+import gobblin.source.workunit.Extract;
+import gobblin.source.workunit.WorkUnit;
+
+
+/**
+ * An implementation of {@link Source} that uses a Hadoop {@link FileInputFormat} to get a {@link FileSplit}
+ * per {@link Extractor} return by {@link #getExtractor(WorkUnitState)} and a {@link RecordReader} to read
+ * the {@link FileSplit}.
+ *
+ * <p>
+ *   This class can read either keys of type {@link #<K>} or values of type {@link #<V>} supported by the
+ *   given {@link FileInputFormat}, through the property {@link #FILE_INPUT_READ_KEYS_KEY}. It will read keys
+ *   if the property is set to {@code true}, otherwise it will read values. By default, it will read values
+ *   through the given {@link FileInputFormat}.
+ * </p>
+ *
+ * <p>
+ *   A concrete implementation of this class should implement {@link #getFileInputFormat(State, Configuration)}
+ *   and {@link #getExtractor(WorkUnitState, RecordReader, FileSplit, boolean)}, which returns a
+ *   {@link HadoopFileInputExtractor} that needs an concrete implementation.
+ * </p>
+ *
+ * @param <S> output schema type
+ * @param <D> output data record type
+ * @param <K> key type expected by the {@link FileInputFormat}
+ * @param <V> value type expected by the {@link FileInputFormat}
+ *
+ * @author ynli
+ */
+public abstract class HadoopFileInputSource<S, D, K, V> implements Source<S, D> {
+
+  private static final String HADOOP_SOURCE_KEY_PREFIX = "source.hadoop.";
+  public static final String FILE_INPUT_FORMAT_CLASS_KEY = HADOOP_SOURCE_KEY_PREFIX + "file.input.format.class";
+  public static final String FILE_SPLITS_DESIRED_KEY = HADOOP_SOURCE_KEY_PREFIX + "file.splits.desired";
+  public static final int DEFAULT_FILE_SPLITS_DESIRED = 1;
+  public static final String FILE_INPUT_PATHS_KEY = HADOOP_SOURCE_KEY_PREFIX + "file.input.paths";
+  public static final String FILE_INPUT_READ_KEYS_KEY = HADOOP_SOURCE_KEY_PREFIX + "file.read.keys";
+  public static final boolean DEFAULT_FILE_INPUT_READ_KEYS = false;
+  static final String FILE_SPLIT_BYTES_STRING_KEY = HADOOP_SOURCE_KEY_PREFIX + "file.split.bytes.string";
```

Comment:
This is only used internally and is not available for the users to set, and it is not used outside the package.